src/test/resources/stela/ft2297/stela_physical_parameters.txt
# $Id: stela_physical_constants.properties 3342 2012-09-19 09:19:39Z ruch $
# =============================================================
# Copyright (c) CNES 2010
# This software is part of STELA, a CNES tool for long term 
# orbit propagation. This source file is licensed as described
# in the file LICENCE which is part of this distribution
# =============================================================

# One Astronomical Unit (m)
astroUnit = 1.49598022291E+11
# Sun pressure for 1 Astro Unit (N/m2)
C0 = 0.45605E-05
# Standard gravitational parameter of the Sun (m3/s2)
muSun = 1.32712440018E20
# Standard gravitational parameter of the Moon (m3/s2)
muMoon = 4.902777900000E12

# Earth Radius (reference semi major axis of Earth ellipsoid) (m)
earthRadius = 6378136.46
# Flattening of the Earth inverse
earthFlatteningInverse = 0.29825765000000E+03
# Standard gravitational parameter of the Earth (m3/s2)
mu = 398600441449820.000
# Zonal gravitational harmonics [Not normalized]
J2 = 0.10826261107488972E-02
J3 = -0.25361508420093188E-05
J4 = -0.16193635251753600E-05
J5 = -0.22289177538077055E-06
J6 = 0.54030052565874444E-06
J7 = -0.36062769005142777E-06
J8 = -0.20807048979064489E-06
J9 = -0.11423641514139811E-06
J10 = -0.23262031558643691E-06
J11 = 0.24092047459748041E-06
J12 = -0.21091044168889998E-06
J13 = -0.21967959574907912E-06
J14 = 0.16882822277732491E-06
J15 = 0.15917867232150533E-09
# Tesseral gravitational harmonics (DONT'T MODIFY THIS COMMENT LINE)
# n m Cnm Snm (n: degree, m: order) [Not normalized]
 2 1 -2.1025792816625992E-10 1.7104172399873343E-9
 2 2 0.0000015745739485232881 -9.038746139036511E-7
 3 1 0.0000021929800852139526 2.6799420442400734E-7
 3 2 3.0905798710326173E-7 -2.11430029406192E-7
 3 3 1.0057358350337047E-7 1.9719819163898114E-7
 4 1 -5.088198810505828E-7 -4.489825492849292E-7
 4 2 7.835159890434121E-8 1.4815064519712534E-7
 4 3 5.921930521502938E-8 -1.2011416432136911E-8
 4 4 -3.9827508652860405E-9 6.524866960150961E-9
 5 1 -5.351599070359425E-8 -8.055895650107465E-8
 5 2 1.055394525675021E-7 -5.230661752681463E-8
 5 3 -1.4915383439989012E-8 -7.091902054796334E-9
 5 4 -2.298899942584298E-9 3.8749972947537444E-10
 5 5 4.304528494011423E-10 -1.648060639740342E-9
 6 1 -5.9549898706351754E-8 2.0496940107799665E-8
 6 2 6.1480927878696425E-9 -4.650458960269791E-8
 6 3 1.1933723115276797E-9 1.8314278334243898E-10
 6 4 -3.2718088755372194E-10 -1.7844098676677768E-9
 6 5 -2.1559699817590804E-10 -4.329492145972659E-10
 6 6 2.2016696767673454E-12 -5.528985322251776E-11
 7 1 2.0457923248388527E-7 6.88682995756525E-8
 7 2 3.292364560085518E-8 9.248821099627783E-9
 7 3 3.5102784410909045E-9 -3.072920661611265E-9
 7 4 -5.844357663167672E-10 -2.635955793077894E-10
 7 5 6.654837188226625E-13 6.4013288635931995E-12
 7 6 -2.4901999588117153E-11 1.053660856478208E-11
 7 7 2.88239698529975E-14 4.4955103951179103E-13
 8 1 1.5892288027175742E-8 4.090015323990022E-8
 8 2 6.421279048909481E-9 5.372278285077551E-9
 8 3 -1.974400859863398E-10 -8.735677968943683E-10
 8 4 -3.17965764411949E-10 9.179890009623351E-11
 8 5 -4.624181475256672E-12 1.6102134670581363E-11
 8 6 -1.8439223569961997E-12 8.635581688101888E-12
 8 7 3.42512486000411E-13 3.82007363068803E-13
 8 8 -1.5809575111730649E-13 1.5370663197244657E-13
 9 1 9.408246013892031E-8 1.5091961561556928E-8
 9 2 1.477201132156169E-9 -2.2386264338882925E-9
 9 3 -1.1954504730864885E-9 -5.477835503228382E-10
 9 4 -7.194133980876692E-12 1.7357807487673435E-11
 9 5 -1.7230632882159018E-12 -5.5247691627897535E-12
 9 6 8.284497632420835E-13 2.9465678090271723E-12
 9 7 -2.2469250278321957E-13 -1.8427199965094786E-13
 9 8 6.149199302876095E-14 -9.864167785359605E-16
 9 9 -3.6797128652783325E-15 7.4563151874383E-15
 10 1 5.137202089295236E-8 -8.124970654758427E-8
 10 2 -5.383279501080332E-9 -3.077831530236937E-9
 10 3 -3.881396049427174E-11 -8.889323061811408E-10
 10 4 -5.0678758464796205E-11 -4.665729379172762E-11
 10 5 -3.0966951724304636E-12 -3.1115051023497704E-12
 10 6 -2.6065918788963095E-13 -5.514374729981028E-13
 10 7 7.008712605311193E-15 -2.53350517115557E-15
 10 8 4.65729057991267E-15 -1.0513742529526015E-14
 10 9 2.3314522942184254E-15 -7.042951116222382E-16
 10 10 4.181576626429528E-16 -9.847542629854147E-17
 11 1 6.380643808550208E-9 -1.8005833519140143E-8
 11 2 1.0446959637502571E-9 -5.0354454096531595E-9
 11 3 -1.5726373916146705E-10 -6.976292854118988E-10
 11 4 -1.6541809573276944E-11 -2.725810549164115E-11
 11 5 1.5346003895938964E-12 2.0052774450705636E-12
 11 6 -3.928622868043544E-15 1.3352851608421763E-13
 11 7 1.827232794921838E-15 -3.728568543837398E-14
 11 8 -2.958162397750247E-16 1.1697640909424142E-15
 11 9 -1.9212232676476857E-16 2.5852056729812463E-16
 11 10 -4.9572480178693956E-17 -1.7521670480850313E-17
 11 11 9.411501045521883E-18 -1.4074122015370588E-17
 12 1 -2.936030352465663E-8 -2.511908733886396E-8
 12 2 3.9697182095679425E-10 1.4723035593688373E-9
 12 3 1.491919903202477E-10 7.920452077202344E-11
 12 4 -2.1039246106768687E-11 1.294947987062766E-12
 12 5 8.631293021100104E-13 1.7811702892949696E-13
 12 6 6.716139847540617E-15 9.272774074329476E-14
 12 7 -4.197702084612098E-15 7.85900508041517E-15
 12 8 -5.776786349634614E-16 3.8362928053019456E-16
 12 9 1.0205105253703644E-16 6.039800200344377E-17
 12 10 -1.8377959954296708E-18 9.284895039955034E-18
 12 11 4.974002004250529E-19 -2.826231082078482E-19
 12 12 -2.1684545854747638E-20 -9.992190435687506E-20
 13 1 -2.469887147582144E-8 2.3369152683165627E-8
 13 2 2.2708450274834393E-9 -2.703473125447463E-9
 13 3 -5.446878165120304E-11 3.045710188533446E-10
 13 4 -2.33918574361176E-13 -2.266674861738015E-12
 13 5 1.049651102526581E-12 1.2288469558819753E-12
 13 6 -5.3086476598799304E-14 -7.800074120444007E-15
 13 7 4.954766167727102E-16 -8.710662280969378E-16
 13 8 -1.1559849779671794E-16 -1.1504677168249936E-16
 13 9 2.6425247440671187E-17 4.878279328755077E-17
 13 10 4.5760790833864766E-18 -4.140692095856307E-18
 13 11 -5.826889061661045E-19 -6.13805161693112E-20
 13 12 -5.821703175641016E-20 1.6414522702312215E-19
 13 13 -2.239392588006351E-20 2.4926120080140657E-20
 14 1 -1.2002504943578748E-8 1.679968849933857E-8
 14 2 -1.0660574806240502E-9 -2.4560770861709937E-10
 14 3 8.900708426273969E-11 6.517877024431843E-11
 14 4 6.523001465890467E-13 -4.102464972485365E-12
 14 5 3.601161931270408E-13 -2.038623602675347E-13
 14 6 -1.8027044494816037E-14 2.5514865616722193E-15
 14 7 2.7994521091837548E-15 -1.7130748433324988E-16
 14 8 -2.0927468927023665E-16 -9.602042326987475E-17
 14 9 1.6284847897332773E-17 1.45061662045022E-17
 14 10 1.8461775892632966E-18 -4.7304224987779565E-20
 14 11 7.417115463726905E-20 -1.8534043113594384E-19
 14 12 4.5210289010574914E-21 -1.6722804788855436E-20
 14 13 2.3511121503059216E-21 3.2937242200057994E-21
 14 14 -7.158702781109039E-22 -6.612402409285407E-23
 15 1 1.8360483470624351E-9 3.3108259772073503E-9
 15 2 -7.013874768943366E-10 -8.039532724106642E-10
 15 3 1.0932170910759611E-10 3.926120741786914E-11
 15 4 -6.062312628426009E-12 5.106872643130872E-13
 15 5 1.318585376067683E-13 9.673815110180314E-14
 15 6 2.2626035944037123E-14 -2.5466478667335476E-14
 15 7 2.7349101521248332E-15 2.5151497680509924E-16
 15 8 -1.0833049077506754E-16 7.971711833135427E-17
 15 9 3.473458326662383E-18 1.0226017379264551E-17
 15 10 2.2556234275928374E-19 3.1662495996782735E-19
 15 11 -2.0833592862961254E-21 3.584488211990715E-20
 15 12 -5.9559389535427494E-21 2.9182906599515126E-21
 15 13 -5.724425804883806E-22 -9.215284294627028E-23
 15 14 1.383767576113049E-23 -6.465724002936946E-23
 15 15 -9.19277228313109E-24 -2.259677232456533E-24

src/main/java/fr/cnes/sirius/patrius/forces/atmospheres/MSIS2000/Flags.java
/**
 * 
 * Copyright 2011-2022 CNES
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 * 
 * @history 03/07/2012
 *
 * HISTORY
 * VERSION:4.10:DM:DM-3185:03/11/2022:[PATRIUS] Decoupage de Patrius en vue de la mise a disposition dans GitHub
 * VERSION:4.9:FA:FA-3128:10/05/2022:[PATRIUS] Historique des modifications et Copyrights 
 * VERSION:4.3:DM:DM-2097:15/05/2019: Mise en conformite du code avec le nouveau standard de codage DYNVOL
 * VERSION::FA:178:06/01/2013:log id format corrected
 * VERSION::DM:130:08/10/2013:MSIS2000 model update
 * END-HISTORY
 */
package fr.cnes.sirius.patrius.forces.atmospheres.MSIS2000;

import java.io.Serializable;

/**
 * 
 * Class Flags
 * Switches: to turn on and off particular variations use these switches.
 * 0 is off, 1 is on, and 2 is main effects off but cross terms on.
 * 
 * Standard values are 0 for switch 0 and 1 for switches 1 to 23. The
 * array "switches" needs to be set accordingly by the calling program.
 * The arrays sw and swc are set internally.
 * 
 * switches[i]:
 * i - explanation
 * -----------------
 * 0 - output in centimeters instead of meters
 * 1 - F10.7 effect on mean
 * 2 - time independent
 * 3 - symmetrical annual
 * 4 - symmetrical semiannual
 * 5 - asymmetrical annual
 * 6 - asymmetrical semiannual
 * 7 - diurnal
 * 8 - semidiurnal
 * 9 - daily ap [when this is set to -1 (!) the pointer
 * ap_a in struct nrlmsise_input must
 * point to a struct ap_array]
 * 10 - all UT/long effects
 * 11 - longitudinal
 * 12 - UT and mixed UT/long
 * 13 - mixed AP/UT/LONG
 * 14 - terdiurnal
 * 15 - departures from diffusive equilibrium
 * 16 - all TINF var
 * 17 - all TLB var
 * 18 - all TN1 var
 * 19 - all S var
 * 20 - all TN2 var
 * 21 - all NLB var
 * 22 - all TN3 var
 * 23 - turbo scale height var
 * 
 * @concurrency not thread-safe
 * 
 * @author Rami Houdroge
 * 
 * @version $Id: Flags.java 17582 2017-05-10 12:58:16Z bignon $
 * 
 * @since 1.2
 * 
 */
public class Flags implements Serializable {

     /** Serializable UID. */
    private static final long serialVersionUID = -4710968807924335177L;

    /** Flag size. */
    private static final int FLAGSIZE = 24;
    /** Daily Ap index. */
    private static final int DAILY_AP_INDEX = 9;

    /** Switches. */
    private int[] switches = new int[FLAGSIZE];
    /** sw. */
    private final double[] sw = new double[FLAGSIZE];
    /** swc. */
    private final double[] swc = new double[FLAGSIZE];

    /**
     * Return boolean from double.
     * 
     * @param fakeBool
     *        double
     * @return boolean from double
     * @since 1.0
     */
    public boolean bool(final double fakeBool) {
        return (fakeBool == 0 ? false : true);
    }

    /**
     * Getter for a particular element in the sw array.
     * 
     * @param position
     *        position in the array
     * @return the element at the given position
     */
    public double getSw(final int position) {
        return this.sw[position];
    }

    /**
     * Getter for a particular element in the swc array.
     * 
     * @param position
     *        position in the array
     * @return the element at the given position
     */
    public double getSwc(final int position) {
        return this.swc[position];
    }

    // GETTERS & SETTERS
    /**
     * Getter for switches.
     * 
     * @return the switches
     */
    @SuppressWarnings("PMD.MethodReturnsInternalArray")
    public int[] getSwitches() {
        return this.switches;
    }

    /**
     * Setter for a specific element of the switches array.
     * 
     * @param position
     *        position in the array
     * @param value
     *        new value.
     */
    public void setSwitches(final int position, final int value) {
        this.switches[position] = value;
    }

    /**
     * Setter for switches.
     * 
     * @param switches
     *        the switches to set
     */
    public void setSwitches(final int[] switches) {
        this.switches = switches;
    }

    /**
     * Prepare sw and swc.
     */
    public void tselec() {
        // loop on the flags size
        for (int i = FLAGSIZE - 1; i >= 0; i--) {
            // if the current index is equal to the Ap index
            if (i == DAILY_AP_INDEX) {
                this.sw[i] = this.switches[i];
                this.swc[i] = this.switches[i];
            } else {
                // test if the switches value for current index is equal to 1
                if (this.switches[i] == 1) {
                    this.sw[i] = 1;
                } else {
                    this.sw[i] = 0;
                }
                // test if the switches value for current index is positve
                if (this.switches[i] > 0) {
                    this.swc[i] = 1;
                } else {
                    this.swc[i] = 0;
                }
            }
        }
    }
}
